File: JavaScript/1043-Partition Array for Maximum Sum.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function(arr, k) {
    const N = arr.length;
    let dp = new Array(N+1).fill(0);
    
    for (i=0; i<arr.length; i++){
        let currMax = arr[i];
        const limit = Math.min(i+1, k);
        for (let j=0; j<limit; j++){
            currMax = Math.max(currMax, arr[i-j]);
            dp[i+1] = Math.max(dp[i+1], dp[i-j] + currMax*(j+1));
        }
    }
    return dp[N];
};

File: C++/1043-Partition Array for Maximum Sum.cpp
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        int N = arr.size();
        vector<int> dp(N+1, 0);
        
        for (int i=0; i<N; i++) {
            int mx = arr[i];
            for (int kk=0; kk<k; kk++) {
                if (kk > i) break;
                mx = max(mx, arr[i-kk]);
                dp[i+1] = max(dp[i+1], mx*(kk+1) + dp[i-kk]);
            }
        }
        
        return dp[N];
    }
};